### ISSUE TITLE
[no-inner-declarations] Function declaration in TS namespace is warned

### ISSUE THREAD

**Issue by ryym** [eslint/eslint#14821]
**Tell us about your environment**

`npx eslint --env-info`

```
Environment Info:

Node version: v14.17.2
npm version: v6.14.13
Local ESLint version: v7.31.0 (Currently used)
Global ESLint version: Not found
Operating System: darwin 19.6.0
```

**What parser (default, `@babel/eslint-parser`, `@typescript-eslint/parser`, etc.) are you using?**

`@typescript-eslint/parser`

Though the problem is specific to TypeScript, it occurs with the builtin rule [no-inner-declarations][no-inner-declarations] so I filed a issue here instead of the [`typescript-eslint`][typescript-eslint] repository.

[no-inner-declarations]: https://eslint.org/docs/rules/no-inner-declarations
[typescript-eslint]: https://github.com/typescript-eslint/typescript-eslint

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
module.exports = {
  root: true,
  parser: "@typescript-eslint/parser",
  rules: {
    "no-inner-declarations": "error",
  },
};
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```ts
// foo.ts
export namespace Foo {
  function hello(): void {
    console.log("Hello");
  }
}
```

```bash
npx eslint foo.ts 
```

**What did you expect to happen?**

No errors occur.  

[namespaces]: https://www.typescriptlang.org/docs/handbook/namespaces.html
[playground]: https://www.typescriptlang.org/play?#code/KYDwDg9gTgLgBAOwIYFtgGcxIMbDgMQgjgG8AoOOAMwFcFsYBLCBOAC2ABtOIAKASgBccAG4RGAE1IVKcbC3QROwAHQ8A5rwBEACS48t-ANwyAvmVNA

**What actually happened? Please copy-paste the actual, raw output from ESLint.**

An error of `no-inner-declarations` rule occur:

```

/Users/yamada.ryutaro/work/tmp-eslint-issue/foo.ts
  3:3  error  Move function declaration to program root  no-inner-declarations

✖ 1 problem (1 error, 0 warnings)

```

I expect `foo.ts` is valid because a function inside a function does not be warned.

```ts
// bar.ts
(function () {
  function hello(): void {
    console.log("Hello");
  }
})();
```

```bash
$ npx eslint bar.ts && echo 'ok'
ok
```

Since [TypeScript's Namespace][namespaces] has its own scope as well as a function, declaring a function inside a namespace should be valid. A namespace is compiled into a function expression like [this][playground].

But of course we can use `let` / `const` instead in most cases so I don't think this is a critical problem.

**Steps to reproduce this issue:**

1. Initialize the Node.js project using this `package.json`:

    ```json
    {
      "name": "tmp-eslint-issue",
      "private": true,
      "dependencies": {
        "@typescript-eslint/parser": "4.28.4",
        "eslint": "7.31.0",
        "typescript": "4.3.5"
      }
    }
    ```

2. Run `npm install`.
3. Add `foo.ts` described above to the project.
4. Run `npx eslint foo.ts`.

**Are you willing to submit a pull request to fix this bug?**

Yes if possible.

**Comment by mdjermanovic** [comment]
Hi @ryym, thanks for the issue!

This works as expected, per [rule details](https://eslint.org/docs/rules/no-inner-declarations#rule-details):

> This rule requires that function declarations and, optionally, variable declarations be in the root of a program or the body of a function.

In this case, function declaration is in the body of a namespace, and core rules are not aware of typescript semantics and the fact that namespace compiles to a function. The best course of action would be to ask `@typescript-eslint/eslint-plugin` for an [extension rule](https://github.com/typescript-eslint/typescript-eslint/tree/master/packages/eslint-plugin#extension-rules).

**Comment by ryym** [comment]
> core rules are not aware of typescript semantics

> The best course of action would be to ask @typescript-eslint/eslint-plugin for an extension rule.

You're right. I didn't come up with suggesting a new extension rule. Thank you for your advice!
